### PR TITLE
wip: end to end multidomain

### PR DESCRIPTION
**Pull request by OpheliaMiralles**
## Description
Adds end-to-end multidomain model support on top of `feature/multi-domain`: all datasets share one encoder, one hidden latent representation and one decoder, with runtime routing for differently sized regional grids. 
Includes also 

- active-domain training/loss handling and 
- an integration test using two CERRA-derived regions with shared ERA5 context and weighted regional loss.

All based on the logic from the original destination Earth branch https://github.com/destination-earth-digital-twins/anemoi-core/tree/DE376.

## What problem does this change solve?
Right now the multidomain branch (#1107) supports only the sampling mechanism for multiple domains. This code adds the component allowing to use the same encoder for datasets with different shapes and support for decoding using the same decoder for several domains. 

```
decoders:
    "0":
      target_datasets:
        - sg_1
        - sg_2
```

## What issue or task does this change relate to?
#1374 #1107 

***As a contributor to the Anemoi framework, please ensure that your changes include unit tests, updates to any affected dependencies and documentation, and have been tested in a parallel setting  (i.e., with multiple GPUs). As a reviewer, you are also responsible for verifying these aspects and requesting changes if they are not adequately addressed. For guidelines about those please refer to https://anemoi.readthedocs.io/en/latest/***

By opening this pull request, I affirm that all authors agree to the [Contributor License Agreement.](https://github.com/ecmwf/codex/blob/main/Legal/Contributor-License-Agreement.md)

----
📚 Documentation preview 📚: https://anemoi-training--1406.org.readthedocs.build/en/1406/

----
📚 Documentation preview 📚: https://anemoi-graphs--1406.org.readthedocs.build/en/1406/

----
📚 Documentation preview 📚: https://anemoi-models--1406.org.readthedocs.build/en/1406/

### CODE REVIEW

File: training/tests/integration/conftest.py [hunk 1]
```diff
@@ -254,11 +255,77 @@ def multidatasets_config(
...
 def multidomain_config(
-    multidatasets_config: tuple[DictConfig, list[str]],
+    stretched_config: tuple[DictConfig, list[str]],
 ) -> tuple[DictConfig, list[str]]:
-    cfg, urls = multidatasets_config
+    cfg, urls = stretched_config
     use_case_modifications = OmegaConf.load(Path.cwd() / "training/tests/integration/config/test_multidomain.yaml")
+    OmegaConf.set_struct(cfg, False)
     cfg = OmegaConf.merge(cfg, use_case_modifications)
+
```
Comment (dietervdb-meteo) [inline training/tests/integration/conftest.py:262]: Could be good to move some of the below explictly to a default config in `training/src/training/configs` and then base the test off of that. will provide a more human readable example of a multi-domain config.

Comment (OpheliaMiralles) [reply training/tests/integration/conftest.py:262]: Agreed! Will do

File: models/src/anemoi/models/models/base.py [hunk 2]
```diff
@@ -71,6 +74,22 @@ def __init__(
 
         self.dataset_names = list(data_indices.keys())
         self._graph_name_hidden = model_config.model.model.hidden_nodes_name
+        self._multiple_hidden_meshes = isinstance(self._graph_name_hidden, (dict, DictConfig))
+        if isinstance(self._graph_name_hidden, (dict, DictConfig)):
+            self.dataset2hidden = dict(self._graph_name_hidden)
```
Comment (dietervdb-meteo) [inline models/src/anemoi/models/models/base.py:79]: I think it would be good to collect all asserts in a helper function, something like `self._validate` or something.

File: models/src/anemoi/models/models/base.py [hunk 3]
```diff
@@ -39,6 +39,9 @@
 class BaseGraphModel(nn.Module):
     """Message passing graph neural network."""
 
+    supports_shared_encoder_decoder = False
```
Comment (dietervdb-meteo) [inline models/src/anemoi/models/models/base.py:42]: Maybe remove this and only add `supports_shared_encoder_decoder = False` to the classes that don't support it? Then check with `self.getattribute('supports_shared_encoder_decoder', True)`, that way, only classes that don't support it need extra code, and then as we bring support to more and more classes we can remove code, rather than updating an attribute that becomes meaningless.

**Comment by einrone** [comment]
As discussed yesterday, I can enable multiple hidden meshes within this PR.

**Comment by dietervdb-meteo** [comment]
Hi @einrone , [c38efa8](https://github.com/ecmwf/anemoi-core/pull/1406/commits/c38efa82a0331b79a89aa16018ee4cbcc553ccab) already takes care of this. So we should review and finetune what is there now.

**Comment by dietervdb-meteo** [comment]
Integration tests appart from accuracy pass. The accuracy issue lives on main: https://github.com/ecmwf/anemoi-core/actions/runs/35847912233. So we are good to merge.
